### ISSUE TITLE
[<Ray component: Data>] FileBasedDatasource's `supports_distributed_reads` attribute is getting incorrectly assigned

### ISSUE THREAD

**Issue by danielezhu** [ray-project/ray#42749]
### The Bug
In the line `self._supports_distributed_reads = not _is_local_scheme(paths)` in [ray.data.datasource.file_based_datasource](https://docs.ray.io/en/master/_modules/ray/data/datasource/file_based_datasource.html#FileBasedDatasource), `_is_local_scheme(paths)` is returning `False` instead of `True` even though the `paths` argument passed to `FileBasedDatasource`'s `__init__` method contains only local-scheme paths.

This is because prior to calling `_is_local_scheme(paths)`, the variable `paths` gets reassigned by this line: `paths, self._filesystem = _resolve_paths_and_filesystem(paths, filesystem)`. Based on inspecting `paths` before and after this line, it seems that this line strips the `local://` prefix from the paths in `paths`. As a result, when `_is_local_scheme` gets called, it will not recognize the paths as being local-scheme paths, even though they are.

### Expected Behavior
I expect `_is_local_scheme(paths)` to return `True` if all paths in the `paths` argument of `FileBasedDatasource`'s `__init__` are local-scheme paths.

### How this impacts me
I'm running a multi-node Ray cluster where at the end of my job, I do some post-processing of a file local to the head node. I read that `jsonl` file using `ray.data.read_json` so that I can perform some minor dataset transformations on it.

What I expect is that the read tasks only get run by the head node, since [this code](https://docs.ray.io/en/latest/_modules/ray/data/read_api.html) in `read_datasource` (pasted below for your convenience) should recognize that the datasource does *not* support distributed reads (since I specify that the file is local scheme).

```
if not datasource.supports_distributed_reads:
        ray_remote_args["scheduling_strategy"] = NodeAffinitySchedulingStrategy(
            ray.get_runtime_context().get_node_id(),
            soft=False,
        )
```

However, since `datasource.supports_distributed_reads` is incorrectly set to `True` due to the bug in question, the scheduling strategy ends up being `SPREAD`. I believe this is why I'm running into issues where worker nodes (who don't have that local file) are trying to read the file.

### Versions / Dependencies

Ray version: 2.9.0
Python version: 3.10.13
OS: macOS Sonoma 14.2.1

I do not believe that the other libraries that I'm using are relevant here.

### First, add some print debugging to `file_based_datasource.py`:
```
print(f"paths before resolution is {paths}")
paths, self._filesystem = _resolve_paths_and_filesystem(paths, filesystem)
paths, file_sizes = map(
            list,
            zip(
                *meta_provider.expand_paths(
                    paths,
                    self._filesystem,
                    partitioning,
                    ignore_missing_paths=ignore_missing_paths,
                )
            ),
        )
print(f"paths after resolution is {paths}")
```

```
self._supports_distributed_reads = not _is_local_scheme(paths)
print(f"self._supports_distributed_reads in file_based_datasource.py is {self._supports_distributed_reads}")
```

### Reproduction script

```
import ray.data
ds = ray.data.read_json("local:///path/to/my/file.jsonl")
```
You will now see the results of the print statements added above.

### Issue Severity

High: It blocks me from completing my task.

**Comment by danielezhu** [comment]
One thing that I don't have full clarity on is the difference between setting the scheduling strategy of `ray_remote_args` as is done [here](https://github.com/ray-project/ray/blob/master/python/ray/data/read_api.py#L307-L311), and setting the scheduling strategy in the remote function options as is done [here](https://github.com/ray-project/ray/blob/master/python/ray/data/read_api.py#L341C9-L343).

In the description of this issue, I surmised that the worker nodes are trying to read the file local to the head node because `ray_remote_args["scheduling_strategy"]` is not being set to the `NodeAffinitySchedulingStrategy`. However, [this comment](https://github.com/ray-project/ray/blob/master/python/ray/data/read_api.py#L334) makes it seem like the only thing that matters is the scheduling strategy that gets set in the remote options, when creating `get_datasource_or_legacy_reader`.

**Question 1**: what is the difference between these two uses of the `NodeAffinitySchedulingStrategy`, and which case is actually causing the worker nodes to read the head node's local file?  

Another observation I've made is that even when the remote options for `get_datasource_or_legacy_reader` include `scheduling_strategy == NodeAffinitySchedulingStrategy`, when the line `read_tasks = datasource_or_legacy_reader.get_read_tasks(requested_parallelism)` gets run and I look inside `file_based_datasource.get_read_tasks`, `ctx.scheduling_strategy` is still `SPREAD`, which seems contradictory.

**Question 2**: is the behavior I observed expected?

Perhaps this behavior is fine, since when `read_datasource` creates the block list after calling `get_read_tasks`, it includes the `ray_remote_args`:

```
block_list = LazyBlockList(
        read_tasks,
        read_stage_name=read_stage_name,
        ray_remote_args=ray_remote_args,
        owned_by_consumer=False,
    )
```

Of course, in this hypothetical case, if we rely on `ray_remote_args["scheduling_strategy"]` to define whether the read tasks get run by worker nodes or not, then we'd need  `ray_remote_args["scheduling_strategy"]` to get set correctly to `NodeAffinitySchedulingStrategy`.

**Comment by bveeramani** [comment]
Hey @danielezhu, the original issue you described has been fixed by https://github.com/ray-project/ray/pull/42308.

> However, [this comment](https://github.com/ray-project/ray/blob/master/python/ray/data/read_api.py#L334) makes it seem like the only thing that matters is the scheduling strategy that gets set in the remote options, when creating get_datasource_or_legacy_reader.

The comment is unrelated to `local:///` schemes. It has to do with Ray Client, which is a legacy thing.

> what is the difference between these two uses of the NodeAffinitySchedulingStrategy, and which case is actually causing the worker nodes to read the head node's local file?

The first determines where the files are read from. The second determines where the datasource object is constructed.

[This snippet](https://github.com/ray-project/ray/blob/master/python/ray/data/read_api.py#L307-L311) makes the read tasks run on the head node. That said, I think you need to also construct the datasource on the head node, since that's when we do file resolution.

> ctx.scheduling_strategy is still SPREAD, which seems contradictory.

`ctx.scheduling_strategy` just describes the default strategy. Since we're using a non-default value here, `ctx.scheduling_strategy` isn't relevant.

**Comment by bveeramani** [comment]
Duplicate of https://github.com/ray-project/ray/issues/42261

**Comment by danielezhu** [comment]
Hi @bveeramani , it looks like the fix you introduced in #42308 got reverted sometime later, because if we look at `file_based_datasource.py` from release `2.9.1` or `2.9.2`, the old code is being used instead (i.e. `_resolve_paths_and_filesystem(paths, filesystem)` is being called before `_is_local_scheme(paths)`).

[Version 2.9.1 code](https://github.com/ray-project/ray/blob/cfbf98c315cfb2710c56039a3c96477d196de049/python/ray/data/datasource/file_based_datasource.py)

[Version 2.9.2 code](https://github.com/ray-project/ray/blob/fce7a361807580953364e2da964f9498f3123bf9/python/ray/data/datasource/file_based_datasource.py)

**Comment by bveeramani** [comment]
@danielezhu the code looks correct on master. 

https://github.com/ray-project/ray/blob/e973f2ed034834457e4aa5084dc7c1f8240923a8/python/ray/data/datasource/file_based_datasource.py#L123-L130

We merged https://github.com/ray-project/ray/pull/42308 after the branch cut for the 2.9.x releases, so we'd expect 2.9.1 and 2.9.2 to not contain the fix.
